### DISCUSSION TITLE
Planetscale with Swift iOS/macOS apps

### DISCUSSION THREAD

**Discussion by bhagyash007** [planetscale/discussion#230] [Documentation] [answered]
As a next gen dev I resonate with the whole server less/ SDK philosophy.

I want to use planetscale on a mac app, I'm used to Firebase SDK's so very unfamiliar with docker, kubernetes setup

One way I thought was using apollo in the app -> Hasura -> Database. But I don't think planetscale works there.

What are all the possible ways to use planetscale with an iOS/macOS app ?  A server less function? Dunno how I can listen to realtime updates from that setup.

**Comment by orware** [comment, marked as answer]
Hello @bhagyash007,

If there is a need for storing data centrally off of the user's devices I think the primary recommended approach when building mobile applications, which in this case would also extend to macOS apps, , is to build an API layer that would help to provide authentication and data access to your application's users.

This link has good general information explaining why this is a good thing:
https://stackoverflow.com/questions/59748819/why-do-we-use-rest-to-connect-to-a-database-on-a-mobile-app

And to think of it another way, without an extra API layer it would essentially require the database credentials for your application to be embedded somewhere directly in the code, which could be reverse engineered / discovered, and result in a compromise of your database and data.

With a RESTful API layer created, you'll abstract away any need to create MySQL connections directly from your application, and instead, after your user's authenticate into your API via your application's code, they would be able to interact / retrieve data from your database via API calls that occur over HTTPS.

This does add extra complexity unfortunately, since this is something that needs to be built out separately from the iOS/macOS code. 

Building an API could be a relatively complex undertaking, depending on the needs of the application, but it is also a general best practice to make it easier to support multiple platforms.

To provide a brief answer to your last question, once an API layer is available, the instance of your application running on the user's device would likely initiate HTTPS API calls which would poll for changes as a simple approach to see if the data in the database has new updates that then would get retrieved and displayed within the application for the user. 

Something closer to real-time might be able to be achieved with an approach that uses WebSockets instead, but I don't have enough of a background in that topic to provide good direction here for you. Additionally, gRPC or other similar protocols may also be worth investigating too (https://www.reddit.com/r/grpc/comments/slg8sj/grpc_for_mobile_apps/).

I hope this helps provide some direction for your questions and good luck as the development for your iOS/macOS apps continues!

**Reply by smotched** [reply]
@orware could planetscale in the future provide an easy to use SDK for mobile devs?

**Reply by orware** [reply]
Hi @smotched,

At the moment I don't believe a mobile dev SDK would be something we'd actively pursue, mainly because I don't think having one would get around the issue of direct database connectivity from within a mobile app being undesirable, as detailed above.

I'm not sure though what kind of functionality you may be imagining would be needed in such an SDK, but it wouldn't hurt to hear more about what you may be thinking of in case there's something there that may make sense to consider in the future.
